arch/arm64/src/imx9/imx9_lpuart.c

```diff
@@ -2356,24 +2340,37 @@ static void imx9_txint(struct uart_dev_s *dev, bool enable)
  * Name: imx9_txready
  *
  * Description:
- *   Return true if the transmit register is available to be written to
+ *   Return true if the transmit fifo is available to be written to
  *
  ****************************************************************************/
 
 static bool imx9_txready(struct uart_dev_s *dev)
 {
   struct imx9_uart_s *priv = (struct imx9_uart_s *)dev;
   uint32_t regval;
+  uint32_t fifo_size;
+  uint32_t fifo_count;
+
+  /* Read the fifo size and current fill ratio. Return true if fifo is not
+   * full
+   */
+
+  regval = imx9_serialin(priv, IMX9_LPUART_FIFO_OFFSET);
+  fifo_size = (regval & LPUART_FIFO_TXFIFOSIZE_MASK) >>
+    LPUART_FIFO_TXFIFOSIZE_SHIFT;
+  fifo_size = fifo_size == 0 ? 1 : (1 << (fifo_size + 1));
+  regval = imx9_serialin(priv, IMX9_LPUART_WATER_OFFSET);
+  fifo_count = (regval & LPUART_WATER_TXCOUNT_MASK) >>
+    LPUART_WATER_TXCOUNT_SHIFT;
 
-  regval = imx9_serialin(priv, IMX9_LPUART_STAT_OFFSET);
-  return ((regval & LPUART_STAT_TDRE) != 0);
+  return fifo_count < fifo_size;
 }
 
 /****************************************************************************
  * Name: imx9_txempty
  *
  * Description:
- *   Return true if the transmit reg is empty
+ *   Return true if the transmit fifo is empty
```

Comment:
This changes the behavior. Checking if the FIFO is empty is not the same thing as checking if the transmit shift register is empty.

Reply:
yes